kukemc/I18n.js

```diff
@@ -0,0 +1,1239 @@
+            TEXT = replaceKeyWithType(TEXT, key, key);
+          }
+          break;
+        }
+        case "ads": {
+          if (this.runtime.ext_moreDataTypes) {
+            TEXT = replaceKeyWithType(
+              TEXT,
+              key,
+              this.runtime.ext_moreDataTypes.tempData.value[key]
+            );
+          } else {
+            TEXT = replaceKeyWithType(TEXT, key, key);
+          }
+          break;
+        }
+        case "scr": {
+          const tmp = this.getVariables();
+          TEXT = replaceKeyWithType(TEXT, key, tmp[key]);
+          break;
+        }
+        default: {
+          console.error(
+            `[kukeMcI18n] unknown replaceText argument type`,
+            `type: ${type}, key: ${key}`
+          );
+          TEXT = replaceKeyWithType(TEXT, key, key);
+        }
+      }
+    }
+
+    return TEXT;
+  }
+
+  /**
+   * @param {String} KEY
+   * @return {*}
+   */
+  getExtraData({ KEY }) {
+    return this.initialedI18n[KEY];
+  }
+
+  /**
+   * @param {String} LANG
+   * @return {*}
+   */
+  setLanguage({ LANG }) {
+    this.language = LANG;
+    this.i18n = this.initialedI18n?.locales[LANG];
+  }
+
+  getLanguageForI18n() {
+    return this.language;
+  }
+
+  getLanguageForBrowser() {
+    return navigator.language;
+  }
+
+  getSupportedLanguages() {
+    return Object.keys(this.initialedI18n?.locales);
+  }
+
+  /**
+   * @param {String} LANG
+   * @return {Boolean}
+   */
+  checkLanguageSupport({ LANG }) {
+    return Object.keys(this.initialedI18n?.locales).includes(LANG);
+  }
+
+  /**
+   * @param {Boolean} needToast
+   * @return {Boolean}
+   */
+  canRequest(needToast = true) {
+    const currentTime = Date.now();
+    const result = currentTime - this.lastRequestTimestamp >= 3000;
+
+    if (result) this.lastRequestTimestamp = Date.now();
+    if (!result && needToast)
+      this.runtime.scratchBlocks.utils?.toast(
+        this.formatMessage("kukeMCI18n.tip.frequently")
+      );
+    return result;
+  }
+
+  /**
+   * @param {String} TEXT 源文本
+   * @param {String} LANG1 源语言
+   * @param {String} LANG2 新语言
+   * @return {Promise<unknown>}
+   */
+  translateText({ TEXT, LANG1, LANG2 }) {
+    if (this.runtime.isPlayerOnly) return;
+    return new Promise((resolve, reject) => {
+      if (LANG1 === LANG2) return resolve(TEXT);
+      if (!this.canRequest()) return resolve("");
+      const xhr = new XMLHttpRequest();
+      xhr.open("POST", "https://api-save.kuke.ink/api/translation", true);
+      xhr.setRequestHeader("Content-Type", "application/json");
+
+      xhr.onload = function () {
+        if (xhr.status === 200) {
+          resolve(xhr.responseText);
+        } else {
+          reject(xhr.statusText);
+        }
+      };
+
+      xhr.onerror = function () {
+        reject(`[kukeMCI18n] request failed`);
+      };
+
+      xhr.send(
+        JSON.stringify({
+          text: TEXT,
+          from: LANG1,
+          to: LANG2,
+        })
+      );
+    });

```

Comment:
XMLHttpRequest 在现代可以使用 fetch 替换。同时请注意 CCW 对外站请求的政策，海外用户的访问速度，以及 CORS。
```suggestion
      if (LANG1 === LANG2) return TEXT;
      if (!this.canRequest()) return "";
      const req = await fetch("https://api-save.kuke.ink/api/translation", {
          method: "POST",
          headers: {
              "Content-Type": "application/json"
          },
          body: JSON.stringify({ text: TEXT, from: LANG1, to: LANG2 })
      });
      return response.ok ? req.text() : req.status;
```